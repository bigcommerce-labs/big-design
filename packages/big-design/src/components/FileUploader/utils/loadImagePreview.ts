export const loadImagePreview = (file: File): Promise<string | null> => {
  if (!file.type.startsWith('image/')) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      const imageUrl = event.target?.result as string;

      resolve(imageUrl);
    };

    reader.onerror = (error) => {
      // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
      reject(error);
    };

    reader.readAsDataURL(file);
  });
};

export const getImagesPreview = async (files: File[]) =>
  files.reduce(async (prevPromise, file, idx) => {
    const prevResult = await prevPromise;
    const imageUrl = await loadImagePreview(file);

    return { ...prevResult, [idx]: imageUrl };
  }, Promise.resolve<Record<number, string | null>>({}));
