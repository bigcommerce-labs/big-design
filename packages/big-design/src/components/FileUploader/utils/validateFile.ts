import { ValidatorConfig } from '../types';

export const validateFileFormat = (file: File | DataTransferItem, fileFormats = '') => {
  const allowedFormats = fileFormats.split(',').map((format) => format.trim());
  const fileExtension = file.type.split('/').pop()?.toLowerCase() ?? '';

  return allowedFormats.some((format) => {
    if (format === '*' || format === '.*') {
      return true;
    } else if (format.startsWith('.')) {
      return format === `.${fileExtension}`;
    } else if (
      format.startsWith('audio/') ||
      format.startsWith('video/') ||
      format.startsWith('image/')
    ) {
      const mediaTypePrefix = format.split('/')[0];

      return file.type.startsWith(`${mediaTypePrefix}/`);
    }

    return true;
  });
};

export const validateFiles = (files: File[], validators: ValidatorConfig[]) =>
  files.flatMap((file, idx) =>
    validators
      .filter(({ validator }) => !validator(file))
      .map(({ message, type }) => ({ file, fileIdx: idx, message, type })),
  );
