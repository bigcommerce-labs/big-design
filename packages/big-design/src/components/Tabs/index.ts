export { Tabs, type TabItem, type TabsProps } from './Tabs';
