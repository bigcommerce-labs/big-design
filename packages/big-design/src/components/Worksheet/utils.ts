import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
