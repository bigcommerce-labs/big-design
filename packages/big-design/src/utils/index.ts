export * from './typedMemo';
export * from './uniqueId';
export * from './warning';
