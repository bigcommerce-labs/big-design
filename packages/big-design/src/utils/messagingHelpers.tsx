import { CheckCircleIcon, ErrorIcon, InfoIcon, WarningIcon } from '@bigcommerce/big-design-icons';
import { ThemeInterface } from '@bigcommerce/big-design-theme';
import React, { ComponentPropsWithoutRef } from 'react';
import { css } from 'styled-components';

import { ButtonProps } from '../components/Button';
import { LinkProps } from '../components/Link';
import { MarginProps } from '../mixins/margins';

export interface SharedMessagingProps extends ComponentPropsWithoutRef<'div'> {
  actions?: MessageAction[];
  header?: string;
  messages: MessageItem[];
  type?: MessagingType;
  onClose?(): void;
}

export type MessagingType = 'success' | 'error' | 'warning' | 'info';

export interface MessageItem {
  text: string;
  link?: MessageLinkItem;
}

export type MessageLinkItem = Pick<LinkProps, 'external' | 'href' | 'target'> & {
  text: string;
};

export interface MessageAction
  extends Omit<ButtonProps, 'children' | 'mobileWidth' | 'variant' | keyof MarginProps> {
  text: string;
  variant?: Exclude<ButtonProps['variant'], 'primary'>;
}

export const getMessagingIcon = (type: MessagingType, condensed?: boolean) => {
  const size = condensed ? 'large' : 'xLarge';

  switch (type) {
    case 'success':
      return <CheckCircleIcon color="success" size={size} />;

    case 'error':
      return <ErrorIcon color="danger" size={size} />;

    case 'warning':
      return <WarningIcon color="warning50" size={size} />;

    case 'info':
      return <InfoIcon color="primary60" size={size} />;
  }
};

export const getBorderStyle = (type: MessagingType, theme: ThemeInterface) => css`
  ${type === 'success' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.success};
  `};

  ${type === 'error' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.danger};
  `};

  ${type === 'warning' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.warning50};
  `};

  ${type === 'info' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.primary60};
  `};
`;

export const getActionVariant = (variant: ButtonProps['variant']) =>
  variant === 'primary' ? 'secondary' : variant;
