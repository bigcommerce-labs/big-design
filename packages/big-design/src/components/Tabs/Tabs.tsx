import React, { ComponentPropsWithoutRef, memo } from 'react';

import { warning } from '../../utils';

import { StyledTab, StyledTabs } from './styled';

export interface TabItem {
  ariaControls?: string;
  id: string;
  title: string;
  disabled?: boolean;
}

export interface TabsProps extends ComponentPropsWithoutRef<'div'> {
  activeTab?: TabItem['id'];
  children?: React.ReactNode;
  items?: TabItem[];
  onTabClick?(tabId: TabItem['id']): void;
}

export const Tabs: React.FC<TabsProps> = memo(
  ({ activeTab, children, className, items = [], onTabClick, style, role, ...props }) => {
    const handleOnTabClick = (event: React.MouseEvent<HTMLElement>) => {
      event.preventDefault();

      const tabId = event.currentTarget.id;

      if (tabId !== activeTab && typeof onTabClick === 'function') {
        onTabClick(tabId);
      }
    };

    const activeItem = activeTab && items[Number(activeTab)];
    const missingAriaControls = items.some((item) => !item.ariaControls);
    const missingFallback =
      activeItem && !document.getElementById(activeItem.ariaControls || `${activeItem.id}-content`);

    if (missingAriaControls || missingFallback) {
      warning(
        'TabItems must have an ariaControls field, otherwise, an element with fallback ID "{tab.id}-content" must exist in the DOM',
      );
    }

    return (
      <StyledTabs {...props} flexDirection="row" role="tablist">
        {items.map(({ ariaControls, id, title, disabled }) => (
          <StyledTab
            activeTab={activeTab}
            aria-controls={ariaControls || `${id}-content`}
            aria-selected={id === activeTab ? 'true' : 'false'}
            disabled={disabled}
            id={id}
            key={id}
            onClick={handleOnTabClick}
            role="tab"
            tabIndex={id === activeTab ? -1 : 0}
          >
            {title}
          </StyledTab>
        ))}
      </StyledTabs>
    );
  },
);

Tabs.displayName = 'Tabs';
