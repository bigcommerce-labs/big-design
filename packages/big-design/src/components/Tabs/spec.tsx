import React from 'react';

import { fireEvent, render } from '@test/utils';

import 'jest-styled-components';

import { TabItem, Tabs } from './Tabs';

const items: TabItem[] = [
  { id: 'tab1', title: 'Tab 1' },
  { id: 'tab2', title: 'Tab 2' },
];

test('render Tabs', () => {
  const { container } = render(<Tabs items={items} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('render Tabs with disabled item', () => {
  const disabledItems: TabItem[] = [...items, { id: 'tab3', title: 'Tab 3', disabled: true }];
  const { container } = render(<Tabs items={disabledItems} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('Tabs has a role', () => {
  const { getByRole } = render(<Tabs items={items} />);

  expect(getByRole('tablist')).toBeInTheDocument();
});

test('Tab items have a role', () => {
  const { getAllByRole } = render(<Tabs items={items} />);

  expect(getAllByRole('tab').length).toBe(2);
});

test('active tab has a tabindex', () => {
  const activeTab = 'tab2';
  const { getAllByRole } = render(<Tabs activeTab={activeTab} items={items} />);

  const tabs = getAllByRole('tab');

  tabs.forEach((tab) => {
    if (tab.id === activeTab) {
      expect(tab.tabIndex).toBe(-1);
    } else {
      expect(tab.tabIndex).toBe(0);
    }
  });
});

test('onTabClick is called', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Tabs items={items} onTabClick={onClick} />);
  const trigger = getByText('Tab 2');

  fireEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();
});

test('active tab changes on tab click', () => {
  const onClick = jest.fn();

  const { getByText, rerender } = render(<Tabs items={items} onTabClick={onClick} />);

  let trigger = getByText('Tab 2');

  fireEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();

  rerender(<Tabs items={items} onTabClick={onClick} />);

  trigger = getByText('Tab 1');

  fireEvent.click(trigger);

  expect(onClick).toHaveBeenCalled();
});

test("clicking a disabled tab doesn't set the active tab", () => {
  let activeTab = 'tab1';
  const disabledItems: TabItem[] = [...items, { id: 'tab3', title: 'Tab 3', disabled: true }];
  const setActiveTab = jest.fn((tabId: string) => (activeTab = tabId));

  const { getByText } = render(<Tabs activeTab={activeTab} items={disabledItems} onTabClick={setActiveTab} />);

  const trigger = getByText('Tab 3');

  fireEvent.click(trigger);

  expect(setActiveTab).not.toHaveBeenCalled();
  expect(activeTab).toBe('tab1');
});

test('does not forward styles', () => {
  const { container } = render(<Tabs className="test" style={{ background: 'red' }} items={items} />);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});
