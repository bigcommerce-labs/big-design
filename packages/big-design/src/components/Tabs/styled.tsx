import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { css } from 'styled-components';

import { StyleableButton } from '../Button/private';

import { TabProps } from './Tabs';

export const StyledTabs = styled.div.attrs({ role: 'tablist' })``;

export const StyledTab = styled(StyleableButton).attrs<TabProps>(props => ({
  role: 'tab',
  tabIndex: props.id === props.activeTab ? -1 : 0,
}))<TabProps>`
  border: none;
  border-bottom: ${({ theme }) => theme.spacing.xxSmall} solid transparent;
  border-bottom-color: ${props => (props.id === props.activeTab ? props.theme.colors.primary40 : 'transparent')};
  border-radius: 0;
  color: ${({ theme }) => theme.colors.secondary70};
  pointer-events: ${props => (props.id === props.activeTab ? 'none' : 'auto')};

  &:active,
  &:hover:not(:active) {
    background-color: transparent;
  }

  &:active {
    box-shadow: none;
  }

  ${props =>
    props.id === props.activeTab &&
    css`
      &:focus {
        box-shadow: none;
      }
    `}
`;

StyledTab.defaultProps = {
  theme: defaultTheme,
  variant: 'subtle',
};

StyledTabs.defaultProps = { theme: defaultTheme };
