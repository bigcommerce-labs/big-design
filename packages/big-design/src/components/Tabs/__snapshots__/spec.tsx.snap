// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Tabs 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  align-content: stretch;
  align-items: stretch;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: flex-start;
  display: flex;
}

.c3 {
  transition: all 150ms ease-out;
  transition-property: background-color,border-color,box-shadow,color;
  align-items: center;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: inline-flex;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  text-decoration: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c3:focus {
  outline: none;
}

.c3[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c3 + .bd-button {
  margin-top: 0.5rem;
}

.c3:active {
  background-color: #DBE3FE;
}

.c3:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c3:hover:not(:active) {
  background-color: #F0F3FF;
}

.c3[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c5 {
  align-content: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c2 {
  overflow: auto;
}

.c4 {
  border: none;
  border-bottom: 0.25rem solid transparent;
  border-bottom-color: transparent;
  border-radius: 0;
  color: #3C64F4;
  pointer-events: auto;
  width: auto;
}

@media (min-width:720px) {
  .c3 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c3 {
    width: auto;
  }
}

<div
  class="c0 c1 c2"
  role="tablist"
>
  <button
    aria-controls="content1"
    aria-selected="false"
    class="c3 c4"
    id="tab1"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 1
    </span>
  </button>
  <button
    aria-controls="content2"
    aria-selected="false"
    class="c3 c4"
    id="tab2"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 2
    </span>
  </button>
</div>
`;

exports[`render Tabs with disabled item 1`] = `
.c0 {
  box-sizing: border-box;
}

.c1 {
  align-content: stretch;
  align-items: stretch;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: flex-start;
  display: flex;
}

.c3 {
  transition: all 150ms ease-out;
  transition-property: background-color,border-color,box-shadow,color;
  align-items: center;
  appearance: none;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #FFFFFF;
  cursor: pointer;
  display: inline-flex;
  flex: none;
  font-size: 1rem;
  font-weight: 400;
  height: 2.25rem;
  justify-content: center;
  line-height: 2rem;
  outline: none;
  padding: 0 1rem;
  position: relative;
  text-align: center;
  text-decoration: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: 100%;
  background-color: transparent;
  border-color: transparent;
  color: #3C64F4;
}

.c3:focus {
  outline: none;
}

.c3[disabled] {
  border-color: #D9DCE9;
  pointer-events: none;
}

.c3 + .bd-button {
  margin-top: 0.5rem;
}

.c3:active {
  background-color: #DBE3FE;
}

.c3:focus {
  box-shadow: 0 0 0 0.25rem #DBE3FE;
}

.c3:hover:not(:active) {
  background-color: #F0F3FF;
}

.c3[disabled] {
  border-color: transparent;
  color: #D9DCE9;
}

.c5 {
  align-content: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

.c2 {
  overflow: auto;
}

.c4 {
  border: none;
  border-bottom: 0.25rem solid transparent;
  border-bottom-color: transparent;
  border-radius: 0;
  color: #3C64F4;
  pointer-events: auto;
  width: auto;
}

@media (min-width:720px) {
  .c3 + .bd-button {
    margin-top: 0;
    margin-left: 0.5rem;
  }
}

@media (min-width:720px) {
  .c3 {
    width: auto;
  }
}

<div
  class="c0 c1 c2"
  role="tablist"
>
  <button
    aria-controls="content1"
    aria-selected="false"
    class="c3 c4"
    id="tab1"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 1
    </span>
  </button>
  <button
    aria-controls="content2"
    aria-selected="false"
    class="c3 c4"
    id="tab2"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 2
    </span>
  </button>
  <button
    aria-controls="tab3-content"
    aria-selected="false"
    class="c3 c4"
    disabled=""
    id="tab3"
    role="tab"
    tabindex="0"
  >
    <span
      class="c5"
    >
      Tab 3
    </span>
  </button>
</div>
`;
